### ISSUE TITLE
'group by' count() soql fails in 9.0

### ISSUE THREAD

**Issue by perbrondum** [mike4aday/SwiftlySalesforce#135]
SELECT  activitydate,  status,  Count(Id) Cnt
            FROM task
            WHERE  isdeleted = false
            GROUP BY activitydate, status

used to work, now it fails with :
Error: keyNotFound(CodingKeys(stringValue: "url", intValue: nil), Swift.DecodingError.Context(codingPath: [CodingKeys(stringValue: "records", intValue: nil), _JSONKey(stringValue: "Index 0", intValue: 0), RecordCodingKey(stringValue: "attributes", intValue: nil)], debugDescription: "No value associated with key CodingKeys(stringValue: \"url\", intValue: nil) (\"url\").", underlyingError: nil))

**Comment by mike4aday** [comment]
@perbrondum would you post your (formatted) code, please?

**Comment by perbrondum** [comment]
```
private func groupedQuery() throws {
    let app = try ConnectedApp()
    let soql = """
        SELECT activitydate, status, Count(Id) Cnt
        FROM task
        WHERE isdeleted = false
        GROUP BY activitydate, status
    """
   // let soql = "SELECT activitydate, status, Count(Id) Cnt FROM task WHERE isdeleted = false GROUP BY activitydate, status"
    let pub: AnyPublisher<QueryResult<SObject>, Error> = app.query(soql: soql)
    pub.sink { completion in
        switch completion {
        case let .failure(error):
            print("Failed to execute grouped Query. Error: \(error)")
        case .finished:
            print("Success query")
            break
        }
    } receiveValue: { (tasks: QueryResult<SObject>) in
        for task in tasks.records {
            print(task.string(forField: "Cnt"))
        }
    }.store(in: &subscriptions)
}
```

**Comment by perbrondum** [comment]
You get the error when you add the GROUP clause.  The count is not really relevant.

**Comment by perbrondum** [comment]
Similar results from use of Max in select list (I assume it's using 'group by' internally).
```
select max(id) 
from account
```

**Comment by mike4aday** [comment]
Thanks @perbrondum - yes, any aggregating function in the query will cause the error. It's my fault - I re-introduced an error in ver. 9 that I had corrected in ver. 7. The current code expects the result to contain an array of records, each with an IDs. But Salesforce returns a list of `AggregateResult` "records" when an aggregating function is used, and there is no ID. I'll fix it ASAP - trying to decide whether to fix with a breaking change (and update to ver 10) or a non-breaking change.

**Comment by perbrondum** [comment]
Is branch V9-fix-135 the fix for the bug?

**Comment by mike4aday** [comment]
@perbrondum it is -- but the fix isn't in there yet. I haven't had time yet to implement the fix. If you are blocked by the bug, I will accelerate the fix and try to get it done this week. I will notify you once ready. Thx

**Comment by perbrondum** [comment]
I have (unfortunately) added some new features (on 9) that I’d love to show
to some people next week, so I’d love to see a fix.

Thanks a lot.

Per

**Comment by perbrondum** [comment]
Thanks a lot for fixing this. Looks like I was able to pass all our tests with this fix.

**Comment by mike4aday** [comment]
@perbrondum good to hear. I didn't have time to test yet - as soon as I have some working unit tests (and figure out what's wrong per issue #137 that you opened) I will merge into main branch.

**Comment by perbrondum** [comment]
Have you decided on a V9 or V10 release for the fix?

**Comment by mike4aday** [comment]
@perbrondum it'll be a minor release, updating version 9. I believe it's ready, but I wanted to do more testing before release. (For version 10, I'm considering a major release, in line with Apple's release of async/await in Swift 5.5.)

**Comment by perbrondum** [comment]
Awesome on both. We’ve started using async/await and can’t wait to see that in 10. 

Per

**Comment by perbrondum** [comment]
Mike - Are you still considering a 9 release for this fix or is it pushed to R10? Thanks.

**Comment by mike4aday** [comment]
@perbrondum - the fix will be in version 10. I'm working on v.10 now, writing tests and plan to release it soon.

**Comment by perbrondum** [comment]
Good to know. Thanks for the update.

Can you share what will be in v10? We’re very interested in anything that
will break connection/SFDC access, Combine/Await? Will salesforce.search
make it back?

Trying to prepare :)

Per B Jakobsen
CEO  Tap2Sales.com
***@***.***

**Comment by mike4aday** [comment]
@perbrondum the major change in v10 will be async/await. I will try to reintroduce SOSL search in v10.0, but if I can't, I'll start working on that for v10.1 as soon as possible. Thanks

**Comment by perbrondum** [comment]
We replaced all SOSL calls w SOQL for now, but would certainly like it back. It’s just more flexible and powerful across objects. 

Thanks 

Per

**Comment by mike4aday** [comment]
Thanks @perbrondum - good for me to know what's useful.

**Comment by mike4aday** [comment]
Fixed in version 10.0.0

**Comment by mike4aday** [comment]
Thank you @perbrondum. I am loving async/await so far - so much simpler to follow, and requires much less code, as you noted. And thanks to the new Actor type, even multiple OAuth requests emanating simultaneously from different places in the code are easily 'bundled' into a single call. (The flip side is that I had to remove, for now at least, the ability to disable the automatic re-authentication -- i.e. no more "allowsLogin" switch in a request. So if you call a Salesforce method and in the unlikely event the refresh token flow fails, the user would be prompted to re-authenticate.)

**Comment by perbrondum** [comment]
Mike — Took me some time to convert, but everything is working (but for
issue #114, which you just commented on, thanks)

Here’s a (very short) list of the issues we encountered. Most of the work
was removing code, learning async/await/Task/Actor (ongoing).

 1. TASK object conflicts with Swift 5.5 Task() and the Swift error message
was not helpful, so that’s 2 hours I’ll never get back :).

 2. When using DispatchQueue.main.async (and asyncAfter) inside functions
like ViewDidLoad(), we get the following 'Sendable' warning.

 Capture of 'self' with non-sendable type ‘xxxViewController' in a
***@***.***` closure

   I’m reading up on Actor/Sendable, eliminating old DispatchQueue code got
rid of most of the warnings but a few I have not resolved and they seem
harmless for now. Not sure if it’s SwiftlySFDC related.

 3. .nextResult is not in V10 (#144). -> Will add code back in when
implemented, not needed for now.

So, again, thanks for a very cool release. It forced us to learn better
coding, eliminated a ton of old ugly code and the simplifications of code
allowed us to optimize the user experience.

Per B Jakobsen
CEO  Tap2Sales.com
***@***.***

**Comment by mike4aday** [comment]
@perbrondum re. the issues you mentioned:
1) Is `TASK` a custom object that you created to represent the Salesforce Task sObject?
2) I'll be interested to know what you learn about this. I'm using Swiftly Salesforce with SwiftUI on a project, so haven't dealt with UIKit issues.
3) I will fix that asap -- please see issue #144 for a workaround in the meantime
Thank you for the feedback.

**Comment by perbrondum** [comment]
1. Yes. Renaming the object fixed the issue
2. I’ll probably do a write-up on how we ended up doing async concurrent and conseq salesforce queries. I’ll pas it along when done. 
We used .zip a lot (loved it) and need that same behavior. 
3. Thanks for the workaround. 

Per
